Fixes #38174: Add remove-brick operation for gluster_volume

devyanikota opened the pull request:
##### SUMMARY
Existing gluster_volume module does not support removal of bricks
from a gluster volume. Also, operation remove-brick needs to be
followed by a commit operation.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
system/gluster_volume

##### ANSIBLE VERSION
```
ansible 2.4.1.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/devyani/work/gdeploy/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /bin/ansible
  python version = 2.7.13 (default, Dec  1 2017, 09:21:53) [GCC 6.4.1 20170727 (Red Hat 6.4.1-1)]
```

##### ADDITIONAL INFORMATION
With the existing gluster_volume module, remove-brick operation fails when trying
to remove a brick from gluster volume.
'remove_bricks' flag handles whether this operation should be performed.
Commit operation is handled based on user's input of the 'force' flag.

The following playbook can be used to perform the remove-brick operation
```
---
- name: Remove a brick
  gluster_volume:
           state=present
           name="{{ volname }}"
           bricks="{{ bricks }}"
           remove_bricks=True
           rebalance=True
           cluster="{{ gluster_hosts }}"
           transport="{{ transport }}"
           replicas="{{ replica_count }}"
           arbiters="{{ arbiter_count }}"
           disperses="{{ disperse_count }}"
           redundancies="{{ redundancy_count }}"
           force="{{ force }}"
```

Bricks are successfully removed and committed from the Gluster volume, followed by
rebalance operation.

Signed-off-by: Devyani Kota <dkota@redhat.com>

Commit:
Fixes #38174: Add remove-brick operation for gluster_volume

Existing gluster_volume module does not support removal of bricks
from a gluster volume. Also, operation remove-brick needs to be
followed by a commit operation.

Signed-off-by: Devyani Kota <dkota@redhat.com>

Commit:
Updates #38174: Add documentation for remove-brick operation

Signed-off-by: Devyani Kota <dkota@redhat.com>

rosmo commented:
shipit

devyanikota commented:
Request to review from Gluster community:
@aravindavk, @amarts, @sac can you please take a look, when you get time.
Thanks.

sac reviewed (approved):
LGTM

Tested on GlusterFS clusters, works fine.

jctanner commented:
bot_status

devyanikota commented:
@rosmo, request to please review. Thanks.

sac commented:
LGTM

alikins commented:
Are there any tests?

devyanikota commented:
@alikins hello, test-playbook? I have added an example in this PR though.
Here is the [playbook](https://paste.fedoraproject.org/paste/0nPKd9QivnH80PXMur~XNg) that I used to test on my 2 node setup.
thanks.

sabose commented on lib/ansible/modules/storage/glusterfs/gluster_volume.py:
```diff
@@ -465,22 +511,28 @@ def main():
                 add_bricks(volume_name, new_bricks, stripes, replicas, force)
                 changed = True
 
+            if removed_bricks:
```
Shouldn't there be a validation on the number of bricks being removed? @sac @devyanikota ?
For instance, if replica count =3, then 3 bricks need to be removed, if removing a sub-volume. Otherwise, replica count needs to be reduced..etc. Is remove-brick supported for plain distribute - shouldn't data migration be triggered?

sac replied:
@sabose we do not validate the bricks since validation is done by gluster. We'd be duplicating the efforts.
Plain distribute is supported, data migration is triggered when we remove the brick. However, commit is tried 12 times with 10 second gap. If the data migration is not complete by then, we give up and it is upto the user to commit. This is because depending on the data migrated it might sometimes take days together to complete.

sabose replied:
Isn't there a difference in how remove-brick is called for replica vs distribute? For distribute volume types - it's remove-brick start followed by remove-brick commit. However if reducing replica count, there's no need for data migration, and hence it's remove-brick force. Does the module account for it? (I'm not sure of the logic of disperse volume types - please check that as well)

devyanikota replied:
I tested the remove-brick operation on various gluster volume types(replicated, distributed, arbiter, disperse).
I was able to successfully remove bricks from the volumes as long as the number of bricks to be removed were of the correct configuration(a multiple of the brick config) of the respective volume type.
@sac @sabose ^^
Resources used:
[Gluster_Admin_guide](https://access.redhat.com/documentation/en-US/Red_Hat_Storage/2.0/html/Administration_Guide/sect-User_Guide-Managing_Volumes-Shrinking.html)
[Gluster docs](https://docs.gluster.org/en/v3/Administrator%20Guide/Managing%20Volumes/#shrinking-volumes)

sabose replied:
ok! @devyanikota so there's no need for the module to append force based on volume types? It's handled  by gluster?

devyanikota replied:
@sabose since its a destructive operation, while discussing the same @sac had suggested mandating force, so that we do not lose any data accidentally. Here the 'force' variable is a check to allow the operation to be executed.
The command for neither of the volumes will include force. Within the retries limit and on the basis of the returned 'status', the 'commit' operation is executed, and the respective gluster message(/error) is returned.

devyanikota commented:
@alikins um, you were reviewing the PR. Can you please drop a comment on the same, so that @bcoca can suggest the next steps to take this forward.
Thank you.

alikins commented:
@devyanikota I'm not a reviewer on this PR. I did ask if there were any tests, but I'm not the maintainer or a reviewer. Afaict maintainer (@rosmo) and reviewers (@sac) say ship it. @sabose seems to have a review open. But don't gate on me, I have no opinions on the PR.

bcoca commented:
bot_status

Commit:
Updates #38174: Add support to reduce cluster configuration.

An exceptional scenario exists where the user might want to
reduce the gluster cluster configuration for replicated
volumes from replica 3 to replica 2 that need to be handled
in a different manner than the generic gluster commands,
where the user is expected to mention the replica count
of the new configuration in order to remove the bricks.

Signed-off-by: Devyani Kota <dkota@redhat.com>

sac commented on lib/ansible/modules/storage/glusterfs/gluster_volume.py:
```diff
@@ -354,6 +366,50 @@ def add_bricks(name, new_bricks, stripe, replica, force):
+    args.extend(removed_bricks)
+    if force:
+        args.append('force')
+    run_gluster(args)
```
When reducing the replica count, the force option is mandatory. This will fail if force is not set.
Also for remove brick can you check there are not pending heals. 

If we are removing a brick from a volume which has pending heals and if that brick had the only good copy, we have data loss.

devyanikota replied:
@sac thanks for pointing it out. I have added checks for self-heal status before executing the remove-brick operation when reducing the cluster configuration.
Thanks.

Commit:
Updates #38174: Add checks for self-heal status prior remove operation.

While reducing the cluster configuration from replica 3
to 2, it needs to be checked for status of self-heal
prior remove-brick operation is executed.

Signed-off-by: Devyani Kota <dkota@redhat.com>

sac commented on lib/ansible/modules/storage/glusterfs/gluster_volume.py:
```diff
@@ -354,6 +366,57 @@ def add_bricks(name, new_bricks, stripe, replica, force):
+
+
+def reduce_config(name, removed_bricks, replicas, force):
+    out = run_gluster(['volume', 'heal', name, 'info', 'summary'])
```
@devyanikota the summary command is added recently. For people running earlier versions of GlusterFS this module will fail. Use volume heal <volname> info and check if number of entries is 0.

sac commented on lib/ansible/modules/storage/glusterfs/gluster_volume.py:
```diff
@@ -354,6 +366,57 @@ def add_bricks(name, new_bricks, stripe, replica, force):
+    summary = out.split("\n")
+    for line in summary:
+        if 'total' in line and line[-2:].strip() != 0:
+            module.fail_json(msg="Operation Aborted, Self-heal in progress.")
```
Nit: Do not capitalize Aborted. Keep the message like:
"Operation aborted, self-heal in progress."

sac commented on lib/ansible/modules/storage/glusterfs/gluster_volume.py:
```diff
@@ -153,6 +153,16 @@
+    cluster:
+      - 10.70.42.85
+    force: true
+  run_once: true
```
Shouldn't an argument replicas be somewhere here?

sac commented on lib/ansible/modules/storage/glusterfs/gluster_volume.py:
```diff
@@ -354,6 +366,57 @@ def add_bricks(name, new_bricks, stripe, replica, force):
+
+
+def reduce_config(name, removed_bricks, replicas, force):
+    out = run_gluster(['volume', 'heal', name, 'info', 'summary'])
```
Summary option is available only in the latest versions. Let us not rely on that, please use heal info and check the count.

devyanikota replied:
@sac Using 'heal info' to extract information on self-heal status, as suggested.

sac commented on lib/ansible/modules/storage/glusterfs/gluster_volume.py:
```diff
@@ -354,6 +366,57 @@ def add_bricks(name, new_bricks, stripe, replica, force):
+    summary = out.split("\n")
+    for line in summary:
+        if 'total' in line and line[-2:].strip() != 0:
+            module.fail_json(msg="Operation Aborted, Self-heal in progress.")
```
Nit - Change the message to:
"Operation aborted, self-heal in progress." removing the capitalisations.

devyanikota replied:
Removed. done.

sac commented on lib/ansible/modules/storage/glusterfs/gluster_volume.py:
```diff
@@ -153,6 +153,16 @@
+    cluster:
+      - 10.70.42.85
+    force: true
+  run_once: true
```
Add an example for removing bricks in case of replica as well. With replica option.

devyanikota replied:
@sac Added an example to reduce cluster configuration.

sabose commented:
@devyanikota thanks for addressing the replica count change in remove_brick

sac commented on lib/ansible/modules/storage/glusterfs/gluster_volume.py:
```diff
@@ -354,6 +377,57 @@ def add_bricks(name, new_bricks, stripe, replica, force):
+    out = run_gluster(['volume', 'heal', name, 'info'])
+    summary = out.split("\n")
+    for line in summary:
+        if 'number' in line and line[-2:].strip() != 0:
```
The output looks like:

Number of entries: 0

And you are matching with number with lower case n. This will not work.
Also you are comparing a string with an int. line[-2:].strip() returns a string you have to cast it to int.

line[-2:] gives last two numbers, let us say the output was '... entries: 100' this logic would return 00. Which will not work. Do a regex match for the number of entries, split everything beyond : to get the actual number of entries remaining.

devyanikota replied:
@sac hey! handled in case of larger numbers as well, rectified. Thanks.

sac reviewed (approved):
Looks good to me. Thanks.

sac commented:
shipit

valentinsiryk commented on lib/ansible/modules/storage/glusterfs/gluster_volume.py:
```diff
@@ -465,22 +539,31 @@ def main():
-            for brick in volumes[volume_name]['bricks']:
-                if brick not in all_bricks:
-                    removed_bricks.append(brick)
+            if not new_bricks and len(all_bricks) < bricks_in_volume:
```
Here is comparison of int and list: `len(all_bricks) < bricks_in_volume`.
`Ansible error: The error was: TypeError: '<' not supported between instances of 'int' and 'list'`

jctanner replied:
@valentinsiryk please file a bug on that.
